### PR TITLE
Allow search fields to provide a query_builder configuration for constructing queries

### PR DESCRIPTION
**Pull request by cbeer**
Building on the pattern from #2366 (and tweaking the facet pattern to mirror search fields, which needed a more generic API...).

### CODE REVIEW

File: lib/blacklight/solr/search_builder_behavior.rb [hunk 1]
```diff
@@ -43,7 +43,7 @@ def add_query_to_solr(solr_parameters)
       solr_parameters[:qt] = blacklight_params[:qt] if blacklight_params[:qt]
 
       if search_field
-        solr_parameters[:qt] = search_field.qt
+        solr_parameters[:qt] = search_field.qt if search_field[:qt]
```
Comment (jcoyne) [inline lib/blacklight/solr/search_builder_behavior.rb:46]: Is there a reason you're using method access and hash access?

File: lib/blacklight/solr/search_builder_behavior.rb [hunk 2]
```diff
@@ -52,30 +52,12 @@ def add_query_to_solr(solr_parameters)
...
-          key.to_s + "=" + solr_param_quote(val, quote: "'")
-        end.join(" ")
-        solr_parameters[:q] = "{!#{local_params}}#{blacklight_params[:q]}"
-
-        ##
-        # Set Solr spellcheck.q to be original user-entered query, without
-        # our local params, otherwise it'll try and spellcheck the local
-        # params!
-        solr_parameters["spellcheck.q"] ||= blacklight_params[:q]
+      if search_field&.query_builder.present?
```
Comment (jcoyne) [inline lib/blacklight/solr/search_builder_behavior.rb:55]: This refactoring is great! 👏

File: lib/blacklight/solr/search_builder_behavior.rb (hunk 1 above)
Comment (cbeer) [reply lib/blacklight/solr/search_builder_behavior.rb]: Nope. Just surrounded by other hash-accessing.

File: lib/blacklight/solr/search_builder_behavior.rb (hunk 2 above)
Comment (cbeer) [reply lib/blacklight/solr/search_builder_behavior.rb:55]: 🤞 ; if nothing else, hopefully it unlocks the "new" JSON query DSL for us.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**